src/computability/DFA.lean

```diff
@@ -43,4 +44,129 @@ def eval := M.eval_from M.start
 def accepts : language α :=
 λ x, M.eval x ∈ M.accept
 
+lemma mem_accepts (x : list α) : x ∈ M.accepts ↔ M.eval_from M.start x ∈ M.accept := by refl
+
+/-- `M.state_path s x` is the states the DFA passes through when evalulating `M.eval_from s x` -/
+def state_path : σ → list α → list σ
+| start [] := [start]
+| start (a :: x) := start :: state_path (M.step start a) x
+
+@[simp] lemma length_state_path (start : σ) (x : list α) :
+  (M.state_path start x).length = x.length + 1 :=
+begin
+  induction x generalizing start,
+  refl,
+  tauto
+end
+
+lemma nth_le_state_path_eq_eval_from_take (start : σ) (x : list α) (n : ℕ)
+(h : n < (M.state_path start x).length) :
+  (M.state_path start x).nth_le n h = M.eval_from start (x.take n) :=
+begin
+  induction n with n ih generalizing start x,
+  { rw [list.take_zero, eval_from, list.foldl],
+    cases x,
+    all_goals
+    { unfold state_path,
+      rw list.nth_le } },
+  { cases x,
+    { unfold state_path at h,
+      rw [list.length_singleton, nat.lt_succ_iff, nat.le_zero_iff] at h,
+      contradiction },
+    unfold state_path,
+    rw [list.take, list.nth_le, eval_from, list.foldl],
+    apply ih }
+end
+
+lemma eval_from_of_append (start : σ) (x y : list α) :
+  M.eval_from start (x ++ y) = M.eval_from (M.eval_from start x) y :=
+begin
+  induction x with a x ih generalizing start,
+  { repeat {rw eval_from},
+    rw [list.nil_append, list.foldl] },
+  { rw [list.cons_append, eval_from, list.foldl],
+    apply ih }
+end
+
+lemma pumping_lemma [fintype σ] (x : list α) (hx : x ∈ M.accepts)
+(hlen : (fintype.card σ + 1) ≤ list.length x) :
+  ∃ a b c,  x = a ++ b ++ c ∧ b ≠ [] ∧ (a ++ b).length ≤ (fintype.card σ + 1) ∧
+  {a} * language.star {b} * {c} ≤ M.accepts :=
+begin
```

Comment:
It would be great to split this huge lemma up, or maybe extract generally useful parts from it.

As for splitting, how about these?
```lean
fintype.card σ + 1 ≤ x.length →
  M.eval_from s x = t →
    ∃ q a b c,
      x = a ++ b ++ c ∧
      a.length + b.length ≤ fintype.card σ + 1 ∧
      b ≠ [] ∧
      M.eval_from s a = q ∧
      M.eval_from q b = q ∧
      M.eval_from q c = t
```

```lean
M.eval_from q x = q → M.eval_from q (x^n) = q
```